### PR TITLE
Implement SC_RESIST_PROPERTY_* and icon display for SC_ARMOR_RESIST

### PR DESCRIPTION
**Pull request by Kenpachi2k13**
### Pull Request Prelude

- [x] I have followed [proper Hercules code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed

I was told that the changes made in #2614 have no effect because of a missing implementation of the `SC_RESIST_PROPERTY_*` status changes. Well, that's true. :grin:
I have to admin that I didn't check this. Sorry. But even if they had fully been implemented, affecting 1 element for each status change would have led to problems anyway.
Because the 4 *proof Potions (12118-12121) and Undead Elemental Scroll (12279) share the same status changes and icons, they'd unpredictably influence each other. For example mutual status change negation or icon removal due to different duration.

My implementation of the  `SC_RESIST_PROPERTY_*` status changes is based on the *proof Potions and affects 2 elements each.
I added icon display to `SC_ARMOR_RESIST` and use it in the Undead Elemental Scroll's script again.

Now comes the tricky part. No one could tell me how these items interact with each other in official servers.
Will using Undead Elemental Scroll cancel all *proof Potion effects?
Will using a *proof Potion be canceled if the Undead Elemental Scroll's effect is active or will it cancel the scroll's effect partially or even completely? Who knows...
In commit 686c575 I added dynamic icon applying/removing/re-applying to have each element's icon is only displayed if its corresponding resistance is greater than 0. For the calculation only values of `SC_RESIST_PROPERTY_*` and `SC_ARMOR_RESIST` are used. Other item or skill effects won't affect it.
I know this solution isn't perfect, so feel free to provide alternatives or official information.

**Issues addressed:** External bug report.

[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

### CODE REVIEW

File: src/map/status.c [hunk 1]
```diff
@@ -8723,40 +8723,186 @@ static int status_change_start_sub(struct block_list *src, struct block_list *bl
...
+				struct status_change_entry *sce_ground = sc->data[SC_RESIST_PROPERTY_GROUND];
+				struct status_change_entry *sce_fire = sc->data[SC_RESIST_PROPERTY_FIRE];
+				struct status_change_entry *sce_wind = sc->data[SC_RESIST_PROPERTY_WIND];
+
+				// Water
+				int sum_water = val1 + ((sce_fire != NULL) ? sce_fire->val2 : 0);
+				bool show_icon = true;
+				if (sce_water != NULL) {
+					sum_water += sce_water->val1;
+					int left = (int)DIFF_TICK(timer->get(sce_water->timer)->tick, timer->gettick());
```
Comment (4144) [inline src/map/status.c:8737]: ``timer->get`` may return NULL. also sce_water->timer can be invalid timer. probably need add checks for timer field and return value of ``timer->get``?

File: src/map/status.c [hunk 2]
```diff
@@ -8723,40 +8723,186 @@ static int status_change_start_sub(struct block_list *src, struct block_list *bl
...
+					clif->sc_end(bl, bl->id, AREA, status->get_sc_icon(SC_RESIST_PROPERTY_WATER));
 				}
 
-				if (val2 > 0) {
+				// Ground
+				int sum_ground = val2 + ((sce_wind != NULL) ? sce_wind->val2 : 0);
+				show_icon = true;
+				if (sce_ground != NULL) {
+					sum_ground += sce_ground->val1;
+					int left = (int)DIFF_TICK(timer->get(sce_ground->timer)->tick, timer->gettick());
```
Comment (4144) [inline src/map/status.c:8754]: ``timer->get`` may return NULL. also sce_ground->timer can be invalid timer. probably need add checks for timer field and return value of ``timer->get``?

File: src/map/status.c [hunk 3]
```diff
@@ -8723,40 +8723,186 @@ static int status_change_start_sub(struct block_list *src, struct block_list *bl
...
+					clif->sc_end(bl, bl->id, AREA, status->get_sc_icon(SC_RESIST_PROPERTY_GROUND));
 				}
 
-				if (val3 > 0) {
+				// Fire
+				int sum_fire = val3 + ((sce_ground != NULL) ? sce_ground->val2 : 0);
+				show_icon = true;
+				if (sce_fire != NULL) {
+					sum_fire += sce_fire->val1;
+					int left = (int)DIFF_TICK(timer->get(sce_fire->timer)->tick, timer->gettick());
```
Comment (4144) [inline src/map/status.c:8771]: ``timer->get`` may return NULL. also sce_fire->timer can be invalid timer. probably need add checks for timer field and return value of ``timer->get``?

File: src/map/status.c [hunk 4]
```diff
@@ -8723,40 +8723,186 @@ static int status_change_start_sub(struct block_list *src, struct block_list *bl
...
+					clif->sc_end(bl, bl->id, AREA, status->get_sc_icon(SC_RESIST_PROPERTY_FIRE));
 				}
 
-				if (val4 > 0) {
+				// Wind
+				int sum_wind = val4 + ((sce_water != NULL) ? sce_water->val2 : 0);
+				show_icon = true;
+				if (sce_wind != NULL) {
+					sum_wind += sce_wind->val1;
+					int left = (int)DIFF_TICK(timer->get(sce_wind->timer)->tick, timer->gettick());
```
Comment (4144) [inline src/map/status.c:8788]: ``timer->get`` may return NULL. also sce_wind->timer can be invalid timer. probably need add checks for timer field and return value of ``timer->get``?

File: src/map/status.c [hunk 5]
```diff
@@ -8723,40 +8723,186 @@ static int status_change_start_sub(struct block_list *src, struct block_list *bl
...
+			case SC_RESIST_PROPERTY_WATER: {
+				struct status_change_entry *sce_all = sc->data[SC_ARMOR_RESIST];
+				struct status_change_entry *sce_fire = sc->data[SC_RESIST_PROPERTY_FIRE];
+				struct status_change_entry *sce_wind = sc->data[SC_RESIST_PROPERTY_WIND];
+
+				// Water
+				int sum_water = val1 + ((sce_fire != NULL) ? sce_fire->val2 : 0);
+				if (sce_all != NULL) {
+					sum_water += sce_all->val1;
+					int left = (int)DIFF_TICK(timer->get(sce_all->timer)->tick, timer->gettick());
```
Comment (4144) [inline src/map/status.c:8811]: ``timer->get`` may return NULL. also sce_all->timer can be invalid timer. probably need add checks for timer field and return value of ``timer->get``?

File: src/map/status.c [hunk 6]
```diff
@@ -8723,40 +8723,186 @@ static int status_change_start_sub(struct block_list *src, struct block_list *bl
...
+			case SC_RESIST_PROPERTY_GROUND: {
+				struct status_change_entry *sce_all = sc->data[SC_ARMOR_RESIST];
+				struct status_change_entry *sce_wind = sc->data[SC_RESIST_PROPERTY_WIND];
+				struct status_change_entry *sce_fire = sc->data[SC_RESIST_PROPERTY_FIRE];
+
+				// Ground
+				int sum_ground = val1 + ((sce_wind != NULL) ? sce_wind->val2 : 0);
+				if (sce_all != NULL) {
+					sum_ground += sce_all->val2;
+					int left = (int)DIFF_TICK(timer->get(sce_all->timer)->tick, timer->gettick());
```
Comment (4144) [inline src/map/status.c:8837]: ``timer->get`` may return NULL. also sce_all->timer can be invalid timer. probably need add checks for timer field and return value of ``timer->get``?

File: src/map/status.c [hunk 7]
```diff
@@ -8723,40 +8723,186 @@ static int status_change_start_sub(struct block_list *src, struct block_list *bl
...
+			case SC_RESIST_PROPERTY_FIRE: {
+				struct status_change_entry *sce_all = sc->data[SC_ARMOR_RESIST];
+				struct status_change_entry *sce_ground = sc->data[SC_RESIST_PROPERTY_GROUND];
+				struct status_change_entry *sce_water = sc->data[SC_RESIST_PROPERTY_WATER];
+
+				// Fire
+				int sum_fire = val1 + ((sce_ground != NULL) ? sce_ground->val2 : 0);
+				if (sce_all != NULL) {
+					sum_fire += sce_all->val3;
+					int left = (int)DIFF_TICK(timer->get(sce_all->timer)->tick, timer->gettick());
```
Comment (4144) [inline src/map/status.c:8863]: ``timer->get`` may return NULL. also sce_all->timer can be invalid timer. probably need add checks for timer field and return value of ``timer->get``?

File: src/map/status.c [hunk 8]
```diff
@@ -8723,40 +8723,186 @@ static int status_change_start_sub(struct block_list *src, struct block_list *bl
...
+			case SC_RESIST_PROPERTY_WIND: {
+				struct status_change_entry *sce_all = sc->data[SC_ARMOR_RESIST];
+				struct status_change_entry *sce_water = sc->data[SC_RESIST_PROPERTY_WATER];
+				struct status_change_entry *sce_ground = sc->data[SC_RESIST_PROPERTY_GROUND];
+
+				// Wind
+				int sum_wind = val1 + ((sce_water != NULL) ? sce_water->val2 : 0);
+				if (sce_all != NULL) {
+					sum_wind += sce_all->val4;
+					int left = (int)DIFF_TICK(timer->get(sce_all->timer)->tick, timer->gettick());
```
Comment (4144) [inline src/map/status.c:8889]: ``timer->get`` may return NULL. also sce_all->timer can be invalid timer. probably need add checks for timer field and return value of ``timer->get``?

File: src/map/status.c [hunk 9]
```diff
@@ -11114,18 +11263,192 @@ static int status_change_end_(struct block_list *bl, enum sc_type type, int tid,
...
+			struct status_change_entry *sce_wind = sc->data[SC_RESIST_PROPERTY_WIND];
 
-			if (sce->val2 > 0)
-				clif->sc_end(bl, bl->id, AREA, status->get_sc_icon(SC_RESIST_PROPERTY_GROUND));
+			// Water
+			clif->sc_end(bl, bl->id, AREA, status->get_sc_icon(SC_RESIST_PROPERTY_WATER));
+			if (sce_water != NULL) {
+				int sum_water = sce_water->val1 + ((sce_fire != NULL) ? sce_fire->val2 : 0);
+				if (sce_water->val1 > 0 && sum_water > 0) {
+					const struct TimerData *td = timer->get(sce_water->timer);
```
Comment (4144) [inline src/map/status.c:11277]: ``timer->get`` may return NULL. also sce_water->timer can be invalid timer. probably need add checks for timer field and return value of ``timer->get``?

File: src/map/status.c [hunk 10]
```diff
@@ -11114,18 +11263,192 @@ static int status_change_end_(struct block_list *bl, enum sc_type type, int tid,
...
+			}
 
-			if (sce->val3 > 0)
-				clif->sc_end(bl, bl->id, AREA, status->get_sc_icon(SC_RESIST_PROPERTY_FIRE));
+			// Ground
+			clif->sc_end(bl, bl->id, AREA, status->get_sc_icon(SC_RESIST_PROPERTY_GROUND));
+			if (sce_ground != NULL) {
+				int sum_ground = sce_ground->val1 + ((sce_wind != NULL) ? sce_wind->val2 : 0);
+				if (sce_ground->val1 > 0 && sum_ground > 0) {
+					const struct TimerData *td = timer->get(sce_ground->timer);
```
Comment (4144) [inline src/map/status.c:11291]: ``timer->get`` may return NULL. also sce_ground->timer can be invalid timer. probably need add checks for timer field and return value of ``timer->get``?

File: src/map/status.c [hunk 11]
```diff
@@ -11114,18 +11263,192 @@ static int status_change_end_(struct block_list *bl, enum sc_type type, int tid,
...
+			}
 
-			if (sce->val4 > 0)
-				clif->sc_end(bl, bl->id, AREA, status->get_sc_icon(SC_RESIST_PROPERTY_WIND));
+			// Fire
+			clif->sc_end(bl, bl->id, AREA, status->get_sc_icon(SC_RESIST_PROPERTY_FIRE));
+			if (sce_fire != NULL) {
+				int sum_fire = sce_fire->val1 + ((sce_ground != NULL) ? sce_ground->val2 : 0);
+				if (sce_fire->val1 > 0 && sum_fire > 0) {
+					const struct TimerData *td = timer->get(sce_fire->timer);
```
Comment (4144) [inline src/map/status.c:11305]: ``timer->get`` may return NULL. also sce_fire->timer can be invalid timer. probably need add checks for timer field and return value of ``timer->get``?

File: src/map/status.c [hunk 12]
```diff
@@ -11114,18 +11263,192 @@ static int status_change_end_(struct block_list *bl, enum sc_type type, int tid,
...
+					clif->status_change_sub(bl, sc_icn, sc_typ, 1, sc_tck, sc_ttl, 0, 0, 0);
+				}
+			}
+
+			// Wind
+			clif->sc_end(bl, bl->id, AREA, status->get_sc_icon(SC_RESIST_PROPERTY_WIND));
+			if (sce_wind != NULL) {
+				int sum_wind = sce_wind->val1 + ((sce_water != NULL) ? sce_water->val2 : 0);
+				if (sce_wind->val1 > 0 && sum_wind > 0) {
+					const struct TimerData *td = timer->get(sce_wind->timer);
```
Comment (4144) [inline src/map/status.c:11319]: ``timer->get`` may return NULL. also sce_wind->timer can be invalid timer. probably need add checks for timer field and return value of ``timer->get``?

File: src/map/status.c [hunk 13]
```diff
@@ -11114,18 +11263,192 @@ static int status_change_end_(struct block_list *bl, enum sc_type type, int tid,
...
+			break;
+		}
+		case SC_RESIST_PROPERTY_WATER:
+			if (sc->data[SC_ARMOR_RESIST] != NULL) {
+				// Water
+				int sum_water = sc->data[SC_ARMOR_RESIST]->val1;
+				if (sc->data[SC_RESIST_PROPERTY_FIRE] != NULL)
+					sum_water += sc->data[SC_RESIST_PROPERTY_FIRE]->val2;
+				if (sc->data[SC_ARMOR_RESIST]->val1 > 0 && sum_water > 0) {
+					const struct TimerData *td = timer->get(sc->data[SC_ARMOR_RESIST]->timer);
```
Comment (4144) [inline src/map/status.c:11337]: ``timer->get`` may return NULL. also sc->data[SC_ARMOR_RESIST]->timer can be invalid timer. probably need add checks for timer field and return value of ``timer->get``?

File: src/map/status.c [hunk 14]
```diff
@@ -11114,18 +11263,192 @@ static int status_change_end_(struct block_list *bl, enum sc_type type, int tid,
...
+					clif->status_change_sub(bl, sc_icn, sc_typ, 1, sc_tck, sc_ttl, 0, 0, 0);
+					remove_icon = false;
+				}
+
+				// Wind
+				int sum_wind = sc->data[SC_ARMOR_RESIST]->val4;
+				if (sc->data[SC_RESIST_PROPERTY_WIND] != NULL)
+					sum_wind += sc->data[SC_RESIST_PROPERTY_WIND]->val1;
+				if (sc->data[SC_ARMOR_RESIST]->val4 > 0 && sum_wind > 0) {
+					const struct TimerData *td = timer->get(sc->data[SC_ARMOR_RESIST]->timer);
```
Comment (4144) [inline src/map/status.c:11351]: ``timer->get`` may return NULL. also sc->data[SC_ARMOR_RESIST]->timer can be invalid timer. probably need add checks for timer field and return value of ``timer->get``?

File: src/map/status.c [hunk 15]
```diff
@@ -11114,18 +11263,192 @@ static int status_change_end_(struct block_list *bl, enum sc_type type, int tid,
...
+
+			break;
+		case SC_RESIST_PROPERTY_GROUND:
+			if (sc->data[SC_ARMOR_RESIST] != NULL) {
+				// Ground
+				int sum_ground = sc->data[SC_ARMOR_RESIST]->val2;
+				if (sc->data[SC_RESIST_PROPERTY_WIND] != NULL)
+					sum_ground += sc->data[SC_RESIST_PROPERTY_WIND]->val2;
+				if (sc->data[SC_ARMOR_RESIST]->val2 > 0 && sum_ground > 0) {
+					const struct TimerData *td = timer->get(sc->data[SC_ARMOR_RESIST]->timer);
```
Comment (4144) [inline src/map/status.c:11368]: ``timer->get`` may return NULL. also sc->data[SC_ARMOR_RESIST]->timer can be invalid timer. probably need add checks for timer field and return value of ``timer->get``?

File: src/map/status.c [hunk 16]
```diff
@@ -11114,18 +11263,192 @@ static int status_change_end_(struct block_list *bl, enum sc_type type, int tid,
...
+					clif->status_change_sub(bl, sc_icn, sc_typ, 1, sc_tck, sc_ttl, 0, 0, 0);
+					remove_icon = false;
+				}
+
+				// Fire
+				int sum_fire = sc->data[SC_ARMOR_RESIST]->val3;
+				if (sc->data[SC_RESIST_PROPERTY_FIRE] != NULL)
+					sum_fire += sc->data[SC_RESIST_PROPERTY_FIRE]->val1;
+				if (sc->data[SC_ARMOR_RESIST]->val3 > 0 && sum_fire > 0) {
+					const struct TimerData *td = timer->get(sc->data[SC_ARMOR_RESIST]->timer);
```
Comment (4144) [inline src/map/status.c:11382]: ``timer->get`` may return NULL. also sc->data[SC_ARMOR_RESIST]->timer can be invalid timer. probably need add checks for timer field and return value of ``timer->get``?

File: src/map/status.c [hunk 17]
```diff
@@ -11114,18 +11263,192 @@ static int status_change_end_(struct block_list *bl, enum sc_type type, int tid,
...
+
+			break;
+		case SC_RESIST_PROPERTY_FIRE:
+			if (sc->data[SC_ARMOR_RESIST] != NULL) {
+				// Fire
+				int sum_fire = sc->data[SC_ARMOR_RESIST]->val3;
+				if (sc->data[SC_RESIST_PROPERTY_GROUND] != NULL)
+					sum_fire += sc->data[SC_RESIST_PROPERTY_GROUND]->val2;
+				if (sc->data[SC_ARMOR_RESIST]->val3 > 0 && sum_fire > 0) {
+					const struct TimerData *td = timer->get(sc->data[SC_ARMOR_RESIST]->timer);
```
Comment (4144) [inline src/map/status.c:11399]: ``timer->get`` may return NULL. also sc->data[SC_ARMOR_RESIST]->timer can be invalid timer. probably need add checks for timer field and return value of ``timer->get``?

File: src/map/status.c [hunk 18]
```diff
@@ -11114,18 +11263,192 @@ static int status_change_end_(struct block_list *bl, enum sc_type type, int tid,
...
+					clif->status_change_sub(bl, sc_icn, sc_typ, 1, sc_tck, sc_ttl, 0, 0, 0);
+					remove_icon = false;
+				}
+
+				// Water
+				int sum_water = sc->data[SC_ARMOR_RESIST]->val1;
+				if (sc->data[SC_RESIST_PROPERTY_WATER] != NULL)
+					sum_water += sc->data[SC_RESIST_PROPERTY_WATER]->val1;
+				if (sc->data[SC_ARMOR_RESIST]->val1 > 0 && sum_water > 0) {
+					const struct TimerData *td = timer->get(sc->data[SC_ARMOR_RESIST]->timer);
```
Comment (4144) [inline src/map/status.c:11413]: ``timer->get`` may return NULL. also sc->data[SC_ARMOR_RESIST]->timer can be invalid timer. probably need add checks for timer field and return value of ``timer->get``?

File: src/map/status.c [hunk 19]
```diff
@@ -11114,18 +11263,192 @@ static int status_change_end_(struct block_list *bl, enum sc_type type, int tid,
...
+
+			break;
+		case SC_RESIST_PROPERTY_WIND:
+			if (sc->data[SC_ARMOR_RESIST] != NULL) {
+				// Wind
+				int sum_wind = sc->data[SC_ARMOR_RESIST]->val4;
+				if (sc->data[SC_RESIST_PROPERTY_WATER] != NULL)
+					sum_wind += sc->data[SC_RESIST_PROPERTY_WATER]->val2;
+				if (sc->data[SC_ARMOR_RESIST]->val4 > 0 && sum_wind > 0) {
+					const struct TimerData *td = timer->get(sc->data[SC_ARMOR_RESIST]->timer);
```
Comment (4144) [inline src/map/status.c:11430]: ``timer->get`` may return NULL. also sc->data[SC_ARMOR_RESIST]->timer can be invalid timer. probably need add checks for timer field and return value of ``timer->get``?

File: src/map/status.c [hunk 20]
```diff
@@ -11114,18 +11263,192 @@ static int status_change_end_(struct block_list *bl, enum sc_type type, int tid,
...
+					clif->status_change_sub(bl, sc_icn, sc_typ, 1, sc_tck, sc_ttl, 0, 0, 0);
+					remove_icon = false;
+				}
+
+				// Ground
+				int sum_ground = sc->data[SC_ARMOR_RESIST]->val2;
+				if (sc->data[SC_RESIST_PROPERTY_GROUND] != NULL)
+					sum_ground += sc->data[SC_RESIST_PROPERTY_GROUND]->val1;
+				if (sc->data[SC_ARMOR_RESIST]->val2 > 0 && sum_ground > 0) {
+					const struct TimerData *td = timer->get(sc->data[SC_ARMOR_RESIST]->timer);
```
Comment (4144) [inline src/map/status.c:11444]: ``timer->get`` may return NULL. also sc->data[SC_ARMOR_RESIST]->timer can be invalid timer. probably need add checks for timer field and return value of ``timer->get``?

**Comment by MishimaHaruna** [comment]
Merging this as is, seen the tricky part you mentioned - we can assume the interaction is correct until anyone reports otherwise

**Pull request merged**